File: Segundo Periodo/AED2/TP03/Pilha.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef struct Jogador{
    int id;
    struct Jogador* next;
    char nome[100];
    int altura;
    int peso;
    char universidade[100];
    int ano_nasc;
    char cidade_nasc[100];
    char estado_nasc[100];
} Jogador;

Jogador* createPlayer(Jogador player){
    Jogador* temp = (Jogador*)malloc(sizeof(Jogador));
    temp->altura = player.altura;
    temp->ano_nasc = player.ano_nasc;
    strcpy(temp->cidade_nasc,player.cidade_nasc);
    strcpy(temp->estado_nasc,player.estado_nasc);
    temp->id = player.id;
    temp->next = NULL;
    strcpy(temp->nome,player.nome);
    temp->peso = player.peso;
    strcpy(temp->universidade, player.universidade);
    return temp;
}

void replaceSubstring(char* str, const char* find, const char* replace) {
    int findLen = strlen(find);
    int replaceLen = strlen(replace);
    char result[1000]; 

    int i, j, index = 0;
    for (i = 0; str[i] != '\0';) {
        int found = 1;

        for (j = 0; j < findLen; j++) {
            if (str[i + j] != find[j]) {
                found = 0;
                break;
            }
        }

        if (found) {
            for (j = 0; j < replaceLen; j++) {
                result[index++] = replace[j];
            }
            i += findLen;
        } else {
            result[index++] = str[i++];
        }
    }

    result[index] = '\0';
    strcpy(str, result);
}

Jogador* add(Jogador* players, int* size, Jogador player) {
    (*size)++;
    players = (Jogador*)realloc(players, (*size) * sizeof(Jogador));
    players[(*size) - 1] = player;
    return players;
}

Jogador* ler(int* size) {
    FILE* file = fopen("/tmp/players.csv", "r");
    if (!file) {
        perror("Erro ao abrir arquivo");
        exit(1);
    }

    Jogador* jogadores = NULL;
    char line[1000];
    fgets(line, sizeof(line), file);

    while (fgets(line, sizeof(line), file)) {
        line[strcspn(line, "\n")] = '\0';
        replaceSubstring(line,",,",",nao informado,");
        if(line[strlen(line)-1] == ','){
            strcat(line,"nao informado");
        }
        char* token = strtok(line, ",");
        int i = 0;
        Jogador jogador;

        while (token != NULL) {
            switch (i) {
                case 0:
                    jogador.id = atoi(token);
                    break;
                case 1:
                    strcpy(jogador.nome, token);
                    break;
                case 2:
                    jogador.altura = atoi(token);
                    break;
                case 3:
                    jogador.peso = atoi(token);
                    break;
                case 4:
                    strcpy(jogador.universidade, token);
                    break;
                case 5:
                    jogador.ano_nasc = atoi(token);
                    break;
                case 6:
                    strcpy(jogador.cidade_nasc, token);
                    break;
                case 7:
                    strcpy(jogador.estado_nasc, token);
                    break;
            }
            token = strtok(NULL, ",");
            i++;
        }
        jogadores = add(jogadores, size, jogador);
    }

    fclose(file);
    return jogadores;
}

void imprimir(Jogador jogador, int i) {
    printf("[%d] ## %s ## %d ## %d ## %d ## %s ## %s ## %s ##\n",
           i, jogador.nome, jogador.altura, jogador.peso,
           jogador.ano_nasc, jogador.universidade, jogador.cidade_nasc,
           jogador.estado_nasc);
}

typedef struct Pilha{
    Jogador* topo;
} Pilha;

Pilha* createPilha(){
    Pilha* pilha = (Pilha*)malloc(sizeof(Pilha));
    pilha->topo = NULL;
    return pilha;
}

void addPilha(Pilha* stack, Jogador player){
    Jogador* temp = createPlayer(player);
    if(temp == NULL){
        printf("Erro ao tentar criar jogador!\n");
    }
    if(stack->topo == NULL){
        stack->topo = temp;
    }else{
        temp->next = stack->topo;
        stack->topo = temp;
    }
}

Jogador* popToPilha(Pilha* pilha){
    Jogador* temp;
    if(pilha->topo == NULL){
        printf("Pilha vazia!");
    }else{
        temp = pilha->topo;
        pilha->topo = pilha->topo->next;
    }
    return temp;
}

void liberJogador(Jogador* player){
    if(player == NULL){
        return;
    }
    liberJogador(player->next);
    free(player);
}

void liberarPilha(Pilha * pilha){
    if(pilha != NULL){
        liberJogador(pilha->topo);
        free(pilha);
    }
}

void printJogador(Jogador* x, int i){
    if(x == NULL){
        return;
    }
    printJogador(x->next, i - 1);
    imprimir(*x,i);
}

int getPilhaLength(Pilha pilha){
    Jogador* player;
    int i = 0;
    for(player = pilha.topo; player != NULL; player = player->next, i++);
    return i;
}

void printarPilha(Pilha* pilha){
    if(pilha->topo == NULL){
        printf("Pilha vazia\n");
    }else{
        int length = getPilhaLength(pilha[0]);
        printJogador(pilha->topo,length-1);
    }
}

 int takeNumber(char * x){
    char num[7];
    int lenX = strlen(x);
    int j = 0, h = 0;
    for(int i = 0; i < lenX; i++){
        if(x[i] == ' '){
            h = ++i;
            i = lenX;
        }
    }
    while(h < lenX){
        num[j] = x[h];
        j++;
        h++;
    }
    return atoi(num);
}

int main() {
    char entrada[100];
    Jogador* players = NULL;
    int size = 0;
    players = ler(&size);
    Pilha* pilha = createPilha();

    while (1) {
        fgets(entrada, sizeof(entrada), stdin);
        entrada[strcspn(entrada, "\n")] = '\0';

        if (strcmp(entrada, "FIM") == 0) {
            break;
        }

        int index = atoi(entrada);
        if (index < size) {
            addPilha(pilha, players[index]);
        }
    }

    int counter = 0;
    scanf("%d",&counter);
    getchar();
    while(counter > 0){
        fgets(entrada, sizeof(entrada), stdin);
        entrada[strcspn(entrada, "\n")] = '\0';
        if(entrada[0] == 'I'){
            int pos = takeNumber(entrada);
            if (pos != -1) {
                addPilha(pilha, players[pos]);
            }
        }else{
            Jogador* temp = popToPilha(pilha);
            printf("(R) %s\n",temp->nome);
            free(temp);
        }
        counter--;
    }

    printarPilha(pilha);
    free(players);
    liberarPilha(pilha);
    return 0;
}

File: Segundo Periodo/AED2/TP03/Pilha.java
import java.io.File;
import java.io.FileNotFoundException;
import java.util.Scanner;

class Jogador {
    private Jogador nextPlayer;
    private Integer id;
    private String nome;
    private Integer altura;
    private Integer peso;
    private String universidade;
    private Integer ano_nasc;
    private String cidade_nasc;
    private String estado_nasc;

    public Jogador(){
        this.id = null;
        this.nome = null;
        this.altura = null;
        this.peso = null;
        this.universidade = null;
        this.ano_nasc = null;
        this.cidade_nasc = null;
        this.estado_nasc = null;
        this.nextPlayer = null;
    }

    public Jogador(int id, String nome, int altura, int peso, String universidade, int ano_nasc, String cidade_nascimento, String estado_nascimento){
        this.id = id;
        this.nome = nome;
        this.altura = altura;
        this.peso = peso;
        this.universidade = universidade;
        this.ano_nasc = ano_nasc;
        this.cidade_nasc = cidade_nascimento;
        this.estado_nasc = estado_nascimento;
        this.nextPlayer = null;
    }

    public void setNextPlayer(Jogador x){
        this.nextPlayer = x;
    }

    public Jogador getNextPlayer(){
        return this.nextPlayer;
    }

    public int getId(){
        return this.id;
    }

    public String getNome(){
        return this.nome;
    }

    public int getAltura(){
        return this.altura;
    }

    public int getPeso(){
        return this.peso;
    }

    public String getUniversidade(){
        return this.universidade;
    }

    public int getAno_Nascimento(){
        return this.ano_nasc;
    }

    public String getCidade_Nascimento(){
        return this.cidade_nasc;
    }

    public String getEstado_Nascimento(){
        return this.estado_nasc;
    }

    public void setId(int id){
        this.id = id;
    }

    public void setNome(String nome){
        this.nome = nome;
    }

    public void setAltura(int altura){
        this.altura = altura;
    }

    public void setPeso(int peso){
        this.peso = peso;
    }

    public void setUniversidade(String universidade){
        this.universidade = universidade;
    }

    public void setAno_Nascimento(int data_nascimento){
        this.ano_nasc = data_nascimento;
    }

    public void setCidade_Nascimento(String cidade_nascimento){
        this.cidade_nasc = cidade_nascimento;
    }

    public void setEstado_Nascimento(String estado_nascimento){
        this.estado_nasc = estado_nascimento;
    }

    public Jogador clone(){
        return new Jogador(id, nome, altura, peso, universidade, ano_nasc, cidade_nasc, estado_nasc);
    }
    @Override
    public String toString(){
        return "## " + nome + " ## " + altura + " ## " + peso + " ## " + ano_nasc + " ## " + universidade + " ## " + cidade_nasc + " ## " + estado_nasc + " ## ";
    }

    public void imprimir(){
        System.out.println(" ## " + nome + " ## " + altura + " ## " + peso + " ## " + ano_nasc + " ## " + universidade + " ## " + cidade_nasc + " ## " + estado_nasc + " ## ");
    }

    public static Jogador[] add(Jogador[] Player,Jogador player){
        Jogador[] players = new Jogador[Player.length+1];
        for(int i = 0; i < Player.length; i++){
            players[i] = Player[i];
        }
        players[Player.length] = player;
        return players;
    } 

    public static Jogador[] ler(){
        File file = new File("players.csv");
        Jogador[] jogadores = new Jogador[0];
        try {
            Scanner scanner = new Scanner(file);
            scanner.nextLine();
            String info = "";
            while (scanner.hasNextLine()) {
                info = scanner.nextLine();
                info = info.replace(",,", ",nao informado,");
                String[] infos = info.split(",");
                if(infos.length != 8){
                    String[] newInfos = new String[infos.length+1];
                    for(int i = 0; i < infos.length; i++){
                        newInfos[i] = infos[i];
                    }
                    newInfos[infos.length] = "nao informado";
                    infos = newInfos;
                }
                
                jogadores = add(jogadores, new Jogador(
                    Integer.parseInt(infos[0]), infos[1], Integer.parseInt(infos[2]),
                    Integer.parseInt(infos[3]), infos[4], Integer.parseInt(infos[5]),
                    infos[6], infos[7]
                ));
            }
            scanner.close();
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        }
        return jogadores;
    }
    
}

public class Pilha {

    Jogador topo;

    public Pilha(){
        topo = null;
    }

    public void add(Jogador player){
        if(topo == null){
            topo = player.clone();
        }else{
            Jogador temp = player.clone();
            temp.setNextPlayer(topo);
            topo = temp;
        }
    }

    public Jogador pop() throws Exception{
        Jogador temp;
        if(topo == null){
            throw new Exception("Pilha vazia!");
        }else{
            temp = topo;
            topo = topo.getNextPlayer();
        }
        return temp;
    }

    public void showJogadores(Jogador temp, int i){
        if(temp == null){
            return;
        }
        showJogadores(temp.getNextPlayer(), i-1);
        System.out.printf("[%d] %s\n", i, temp.toString());
    }

    public void showPilha(){
        int length = countJogadores();
        showJogadores(topo,length-1);
    }

    public int countJogadores(){
        int i = 0;
        for(Jogador temp = topo; temp != null; i++, temp = temp.getNextPlayer());
        return i;
    }

    public static void main(String[] args) {
        String entrada = "";
        Jogador[] play = Jogador.ler();
        Pilha pilha = new Pilha();
        while(!entrada.equals("FIM")){
            entrada = MyIO.readLine();
            if(entrada.equals("FIM")){
                continue;
            }
            pilha.add(play[Integer.parseInt(entrada)]);
        }
        int counter = MyIO.readInt(), i = 0;
        while (i < counter) {
            String[] sep;
            if(i == counter-1){
                break;
            }
            entrada = MyIO.readLine();
            if (entrada.charAt(0) == 'R') {
                try {
                    System.out.println("(R) " + pilha.pop().getNome());
                } catch (Exception e) {
                    e.printStackTrace();
                }
            } else {
                sep = entrada.split(" ");
                pilha.add(play[Integer.parseInt(sep[1])]);
            }
            i++;
        }
        pilha.showPilha();
    }
}